src/address.c

```diff
@@ -0,0 +1,81 @@
+#include "internal.h"
+#include "base58.h"
+#include "ccan/ccan/build_assert/build_assert.h"
+#include <include/wally_address.h>
+#include <include/wally_bip32.h>
+#include <include/wally_crypto.h>
+#include <include/wally_script.h>
+
+int wally_bip32_key_to_address(const struct ext_key *hdkey, uint32_t flags,
+                               uint32_t version, char **output)
+{
+    int ret;
+
+    if (output)
+        *output = NULL;
+
+    if (!hdkey || (version & ~0xff) || (flags & ~0xff) || !output)
+        return WALLY_EINVAL;
+
+    if (!(flags & (WALLY_ADDRESS_TYPE_P2PKH | WALLY_ADDRESS_TYPE_P2SH_P2WPKH)))
+        return WALLY_EINVAL;
+
+    // Catch known incorrect combinations of address type and version:
+    if (
+        (flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0x05) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x00) ||
+        (flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0xC4) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x6F)
+        ) {
+        return WALLY_EINVAL;
+    }
+
+    unsigned char address[HASH160_LEN + 1];
+
+    if (flags == WALLY_ADDRESS_TYPE_P2PKH) {
+        // pub_key_hash = ripemd160(sha256(pubkey)
+        address[0] = (unsigned char) version & 0xff;
+        if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), address + 1, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+
+        ret = wally_base58_from_bytes(address, sizeof(address), BASE58_FLAG_CHECKSUM, output);
+    } else {
+        BUILD_ASSERT(flags == WALLY_ADDRESS_TYPE_P2SH_P2WPKH);
```

Comment:
Should this be a runtime check?

Reply:
This should never happen, because of the above `Catch known incorrect combinations` checks. I don't know if the compiler is smart enough to understand that though. If not, then I should make it a runtime check (but I couldn't get `assert` to work).

Reply:
the comment has more to do with the fact that this isn't a compile time expression and will cause a sizeof evaluatiion of a VLA effectively. I think it needs to be converted to an if expression or removed.

Reply:
This goes way beyond my C comprehension. I'll just drop the `BUILD_ASSERT` then.